src/engine/interop/specification.md

```diff
@@ -0,0 +1,130 @@
+
+#### Returns
+1. `payloadId`: `QUANTITY` - identifier of the payload building process
+
+#### Specification
+
+1. Given provided field value set client software **SHOULD** build the initial version of the payload which has an empty transaction set.
+
+2. Client software **SHOULD** start the process of updating the payload. The strategy of this process is implementation dependent. The default strategy would be to keep the transaction set up-to-date with the state of local mempool.
+
+3. Client software **SHOULD** stop the updating process either by finishing to serve the [**`engine_getPayload`**](#engine_getPayload) call with the same `payloadId` value or when [`SECONDS_PER_SLOT`](https://github.com/ethereum/consensus-specs/blob/dev/specs/phase0/beacon-chain.md#time-parameters-1) (currently set to 12 in the Mainnet configuration) seconds have passed since the point in time identified by the `timestamp` parameter.
+
+4. Client software **MUST** set the payload field values according to the set of parameters passed in the call to this method with exception for the `feeRecipient`. The `coinbase` field value **MAY** stem from what is specified by the `feeRecipient` parameter.
+
+### engine_getPayload
+
+#### Parameters
+1. `payloadId`: `QUANTITY` - Identifier of the payload building process
+
+#### Returns
+`Object|Error` - Either instance of [`ExecutionPayload`](#ExecutionPayload) or an error
+
+#### Specification
+
+1. Given the `payloadId` client software **MUST** respond with the most recent version of the payload that is available in the corresponding building process at the time of receiving the call.
+
+2. The call **MUST** be responded with error code `2` (`Action not allowed`) if a building process identified by the `payloadId` doesn't exist.
+
+3. Client software **MAY** stop the corresponding building process after serving this call.
+
+### engine_executePayload
+
+#### Parameters
+1. `Object` - Instance of [`ExecutionPayload`](#ExecutionPayload)
+
+#### Returns
+`Object` - Response object:
+1. `status`: `String` - the result of the payload execution:
+  - `VALID` - given payload is valid
+  - `INVALID` - given payload is invalid
+  - `SYNCING` - sync process is in progress
+
+#### Specification
+
+1. Client software **MUST** validate the payload according to the execution environment rule set defined in the [EIP-3675](https://eips.ethereum.org/EIPS/eip-3675#specification) and respond with the validation result.
+
+2. Client software **MUST** defer persisting a valid payload until the corresponding [**`engine_consensusValidated`**](#engine_consensusValidated) message deems the payload valid with respect to the proof-of-stake consensus rules.
+
+3. Client software **MUST** discard the payload if it's deemed invalid.
+
+4. The call **MUST** be responded with `SYNCING` status while the sync process is in progress and thus the execution cannot yet be validated.
```

Comment:
I would assume that in this situation the payload is discarded by the execution engine, but this should be made clear.

Reply:
I would defer the decision of whether to keep or discard the payload in this case to the execution client. As if the execution client is being synced then it rather store all the payloads and then prune the storage upon receiving finalized block as current sync proposal depends on finalized checkpoints and in advance it is unknown which one is already finalized or will be finalized soon. Also, execution clients might want to store blocks of the finalized chain and prune the others.

Reply:
I think my concern here is that the caller ends up in limbo at current regarding the state of the payload.  It may be discarded, it may be kept until the node is synced and then either integrated in to the chain or dropped, it may be kept only if the node is nearly synced and then broadcast if integrated in to the chain, _etc_.  Seems like the safest and clearest option would be for the node to discard the payload if it cannot process it immediately.

(Could storing the blocks be a potential DoS vector, where a malicious attacker would flood a syncing node with future payloads the node needs to hold around in case they are used?)

Reply:
Consensus client should verify the beacon block and send the corresponding `consensusValidated` message disregarding the sync status of the execution client. If `consensusValidated` deems the payload as invalid then it must be discarded as it is stated in the specification to the method below. During the sync all non-canonical forks may still be pruned upon receiving `forkchoiceUpdated`. I think execution client should be free to decide how to handle these calls while in a sync mode and rely on consensus client in sorting out invalid blocks.

Reply:
Right, it is valuable to continue to pass these blocks into the execution-engine even during SYNCING because the execution engine (depending on sync method) will often need more current blocks to be able to sufficiently finish sync.

Eventually, the beacon chain will pass in a payload that is `VALID` or `INVALID` instead of `SYNCING` which is a signal that sync is completed and payloads are being fully validated